django/contrib/auth/hashers.py

```diff
@@ -416,35 +453,50 @@ def encode(self, password, salt):
         data = bcrypt.hashpw(password, salt)
         return "%s$%s" % (self.algorithm, data.decode('ascii'))
 
-    def verify(self, password, encoded):
-        algorithm, data = encoded.split('$', 1)
+    def decode(self, encoded):
+        algorithm, empty, algostr, work_factor, data = encoded.split('$', 4)
         assert algorithm == self.algorithm
+        salt, checksum = data[:22], data[22:]
+        return {
+            'algorithm': algorithm,
+            'algostr': algostr,
+            'work_factor': work_factor,
+            'salt': salt,
+            'checksum': checksum,
+        }
+
+    def verify(self, password, encoded):
+        decoded_entry = self.decode(encoded)
+        partial = "".join((decoded_entry['salt'],
```

Comment:
For simplicity I think it would be better to revert this to the previous code, ie simply https://github.com/django/django/blob/69e0d9c553bb55dde8d7d1d479a78bfa7093f406/django/contrib/auth/hashers.py#L425-L427 -- I understand your motivation behind using `decode` here, but simplicity wins especially in security relevant code.

Reply:
I'm fine with reverting the implementation. Should I add the `assert algorithm == self.algorithm` back as well?

Reply:
Yes, if you do not call `decode` the `assert` should be there. The assert is just a `safeguard` that you don't try to decode a hash with the wrong hasher, it's not something that should happen during normal operations anyways. And if one is running python with `-O`, `assert` will be completely ignored. I agree that the assert are not perfect in those cases, but I wouldn't extend the scope of the PR to also change that.

Reply:
Cool. Pushed a pair of commits to revert `verify` and `harden_runtime` for the `BCryptSHA256PasswordHasher`.